### PR TITLE
make printing of typealias env shorter using `{<:T}` syntax, when possible

### LINKED ISSUE #34887: show for UnionAll types prints non-syntactic identifiers for unnamed type variables
Calling `Base.show()` with `UnionAll` types with unnamed type parameters prints those type parameters with non-syntactic generated names:
```julia
julia> Vector{<:Bool}
Array{#s16,1} where #s16<:Bool
```
This of course cannot be evaluated, which can be difficult when trying to understand a very large type that's printed in an error message, for example.

My first thought was that this should be fixed following the same approach as in https://github.com/JuliaLang/julia/pull/32408: we should change these to print via the `var""` syntax, a la:
```julia
julia> Vector{<:Bool}
Array{var"#s16",1} where var"#s16"<:Bool
```

That seems like a fine approach, but then I noticed that if the variable _does_ have a name, the printing mechanism seems to be able to tweak the user-provided name to prevent conflicts with other identifiers in the type! This leads me to wonder if we can use that same smarts to just pick a standard identifier name, rather than a purposefully non-standard identifier as above?
For example, notice how the user-provided name `T` gets renamed to `T1` when there's a conflict:
```julia
julia> NTuple
Tuple{Vararg{T,N}} where T where N

julia> NTuple{N,NTuple} where N
Tuple{Vararg{Tuple{Vararg{T,N}} where T where N,N}} where N

julia> NTuple{1,NTuple{T}} where T
Tuple{Tuple{Vararg{T1,T}} where T1} where T
```

Can we similarly have julia just pick some arbitrary type names like `p1`, `p2`, `pN`... or `P1`, `P2`, ... or `_T1`, `_T2`, etc, rather than the `#s16`, `#s17` type names we're using now? And then if there's a conflict with an actual user-provided name, it can just adjust the name like it already does?
In fact, the existing mechanism already handles conflicts even for the generated names, so this should be a very easy change: just simplifying the generated names to be valid identifiers.
```julia
julia> Vector{<:Vector{<:Bool}}
Array{#s16,1} where #s16<:(Array{#s17,1} where #s17<:Bool)

julia> Vector{<:Vector{var"#s16"}} where var"#s16" <: Bool
Array{#s17,1} where #s17<:Array{#s16,1} where #s16<:Bool

julia> Vector{<:Vector{var"#s17"}} where var"#s17" <: Bool
Array{#s171,1} where #s171<:Array{#s17,1} where #s17<:Bool
```

### PR DESCRIPTION
**Pull request by vtjnash**
Inspired by https://github.com/JuliaLang/julia/pull/35710, including taking the tests (though not any of the implementation) from there. Unlike the original PR, this also prefers the short form (eliding the TypeVar name), even in case where it is not a gensym.
```
julia> struct A{B,C,D,E,F,G} end

julia> A{U, <:Any, T, Int, S, <:T} where {S, T<:S, U<:Integer}
A{<:Integer, <:Any, T, Int64, S, <:T} where {S, T<:S}
```

Github: this closes #35710 and fixes #34887

### CODE REVIEW

File: base/show.jl [hunk 1]
```diff
@@ -597,6 +597,62 @@ function make_typealias(@nospecialize(x::Type))
...
+    end
+    return true
+end
+
+function show_typeparams(io::IO, env::SimpleVector, orig::SimpleVector, wheres::Vector)
+    n = length(env)
+    elide = length(wheres)
+    for i = n:-1:1
+        p = env[i]
+        if p isa TypeVar
```
Comment (stev47) [inline base/show.jl:620]: maybe reduce nesting: `p isa TypeVar || continue`?

File: base/show.jl [hunk 2]
```diff
@@ -597,6 +597,62 @@ function make_typealias(@nospecialize(x::Type))
...
+            if i == n && orig[i] === p && show_can_elide(p, wheres, elide, env, i)
+                n -= 1
+                elide -= 1
+            elseif p.lb === Union{} && show_can_elide(p, wheres, elide, env, i)
+                elide -= 1
+            elseif p.ub === Any && show_can_elide(p, wheres, elide, env, i)
```
Comment (stev47) [inline base/show.jl:626]: can merge with `Union{}` branch

**Review by stev47 — COMMENTED** [review]
Nice, I must have missed the `has_typevar` intrinsic back when I tried to do this. Avoiding additional io context keys using the mutable `wheres` is probably good too. It seems solid from my quick tests.

It would be nice if `Ref{T} where T` and `Ref{X} where X` both printed consistently, though I guess this requires non-trivial changes in how you build `orig` for `show_typeparams`.

File: base/show.jl (hunk 1 above)
Comment (vtjnash) [reply base/show.jl:620]: there is an else clause, so this is the right structure

Comment (vtjnash) [reply base/show.jl:620]: oh, I see this is the other loop. The reason is still the same though (keeping the general structure of the two loops in common)

**Comment by vtjnash** [comment]
I agree those should do the same (as did CI). I've used the new `jl_types_egal` function to handle that now, so let's see how CI fares now.

**Comment by JeffBezanson** [comment]
I think we should put back the gensym heuristic; i.e. still print `where name` for non-gensyms. It's nice to print something egal to the original input when possible, and less likely to break tests that might exist out there.

**Comment by vtjnash** [comment]
Okay, top example is now:
```
julia> struct A{B,C,D,E,F,G} end

julia> A{U, <:Any, T, Int, S, <:T} where {S, T<:S, U<:Integer}
A{U, <:Any, T, Int64, S, <:T} where {S, T<:S, U<:Integer}
```
with added gensym (`#`) test.

**Comment by vtjnash** [comment]
Thinking to merge this tomorrow, since I haven't heard any concerns with this now

**Comment by StefanKarpinski** [comment]
I actually prefer the more aggressively normalized original form. It's easier to see what's happening, IMO. Parameter names aren't significant after all, so eliminating where possible seems good.

**Comment by vtjnash** [comment]
There's some cases were readability was slightly lower. Like hiding parameter names to functions too–they are not a significant part of the call, but they can be useful. However, as Jeff mentioned, it is also slightly more preferable to print an egal copy when feasible. I still think we should merge this now, but partly because we can still easily make that choice later after getting some experience with this.

**Comment by StefanKarpinski** [comment]
Fair enough, although I think we should probably make a decision before 1.7 to minimize the number of disruptive changes to printing. Although otoh maybe it's good to get people to stop testing exact printing of types.

**Comment by vtjnash** [comment]
We've made more drastic changes than this is in 1.5 and 1.6 already, so hopefully that's already having an influence 😬

**Comment by NHDaly** [comment]
Cool, thanks! :)

**Pull request merged**